merging-configmaps-gitops/README.md

```diff
@@ -0,0 +1,84 @@
+# Merging config in a gitops context
+
+Our app platform utilizes configmaps / secrets heavily to pass configuration into apps.
+
+*The following RFC will focus on configmaps but the same problem and solution are also applicable for secrets!*
+
+These configmaps are often either shared between multiple instances of the same app or have shared values between apps.
+In gitops the [intended solution](https://kubectl.docs.kubernetes.io/guides/introduction/kustomize/#2-create-variants-using-overlays) for both of these cases is to use `bases` and `overwrites` which is achieved through `kustomize`.
+
+## Problem Statement
+
+The `data` section of a configmap is a mapping of keys to values.
+Therefore configuration (especially nested configuration) is usually grouped under a single key (e.g. `values`).
+
+It is currently not possible to merge any `overwrite` into a configmap key because the associated value is **always** treated as a string.
+
+Example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    giantswarm.io/managed-by: flux
+  name: flux01-config
+  namespace: org-multi-project
+data:
+  values: |
+    nodePools:
+    - class: default
+      failureDomain: gb-lon-1
+      name: default
+      replicas: 3
+```
+All fields under `values` will be grouped into a simple `string` even though they are valid `yaml`.
+The only way to overwrite would be to overwrite the entire content of the `values` key.
+
+## Context
+
+This problem became most apparent while working to make Cluster API releases more DRY by using bases.
+All parts of our setup currently rely on `app` CRs and their configuration through configmaps.
+
+Full context of the PR can be found here: https://github.com/giantswarm/workload-clusters-fleet/pull/130
+
+## Solution Proposal
+
+Our app platform offers a layerd configuration approach already which we can utilize to enable us here.
+
+An app CR has two fields where configuration can be supplied:
+```yaml
+apiVersion: application.giantswarm.io/v1alpha1
+kind: App
+metadata:
+  name: something
+  namespace: org-some
+spec:
+  config:
+    configMap:
+      name: flux01-default-apps-config
+      namespace: org-some
+  userConfig:
+    configMap:
+      name: flux01-userconfig
+      namespace: org-some
+```
+This functionality means that values in `config` will be overlaid by values from `userConfig`.
+So far we have only been using this functionality to supply Giant Swarm offered configuration through `config` and encouraged end users to only use the `userConfig` field.
+
+The proposal is to use `config` for cases in gitops where we can not merge configmaps through `kustomize`.
+From preliminary testing this works without any issues **but** is a change to how we treated this feature so far in the app platform.
+
+### Pros
+- Easy to just use
+- Easy to explain to customers
+
+### Cons
+- Change in the intended design in app platform
```

Comment:
tbh, having been part of the design team on this, I don't see it as a big change, it was intended to model 2 levels of config just like an overlay. Our use case for it was having cluster level config in the one level and user config in the 2nd. In the case of MC level usage the cluster level is "free" and we can re-use it for sth else like you are proposing here, and which I've also been hinting at before in Slack.

Reply:
Are there cases (such as e.g. `prometheus`) where we already use this for WC apps? Or does this need to be checked on a case by case basis?

Reply:
I agree with Puja this is just using the design for a different use case.

However this will only work for MC apps. For WC apps we always default the cluster values configmap for use cases like setting the base domain. https://docs.giantswarm.io/app-platform/defaulting-validation/#defaulting

This could be changed but changing it for existing WCs apps it not straightforward.

Reply:
But would this mean if we supply a custom `config`, we just have to make sure the required fields are set? E.g. the `base domain`?

Reply:
For MC apps including the default apps bundle we're fine because `.spec.config` is not currently used.

For the child default apps in the WC the config can be passed via `.spec.userConfig` but `.spec.config` is in use.

For WC apps we could set `.spec.config` so it's not defaulted but any values that are in use like the base domain would need to be present.